src/GeniusYield/Providers/LiteChainIndex.hs

```diff
@@ -0,0 +1,205 @@
+    closeLCIClient,
+    lciWaitUntilSlot,
+    lciLookupDatum,
+    lciGetCurrentSlot,
+    lciStats,
+) where
+
+import           GeniusYield.Imports
+import           GeniusYield.Types
+
+import qualified Cardano.Api                  as Api
+import qualified Cardano.Api.ChainSync.Client as Api.Sync
+import qualified Control.Concurrent.Async     as Async
+import qualified Control.Concurrent.STM       as STM
+import qualified Data.Map.Strict              as Map
+
+-- | A very simple chain index client, which only maintains a datum hashes.
+--
+data LCIClient = LCIClient
+    (Async.Async ())
+    (STM.TVar Api.SlotNo)
+    (STM.TVar (Map (Api.Hash Api.ScriptData) Api.HashableScriptData))
+
+withLCIClient
+    :: Api.LocalNodeConnectInfo
+    -> [Api.ChainPoint]                             -- ^ resume points
+    -> (LCIClient -> IO r)
+    -> IO r
+withLCIClient info resumePoints kont = do
+    slotVar <- STM.newTVarIO $ Api.SlotNo 0
+    dataVar <- STM.newTVarIO Map.empty
+
+    let cb = chainSyncCallback slotVar dataVar
+
+    withChainSync info resumePoints cb $ \a -> kont $ LCIClient
+        a
+        slotVar
+        dataVar
+
+-- | Create new 'LCIClient'.
+--
+-- Use 'withLCIClient' if possible.
+newLCIClient
+    :: Api.LocalNodeConnectInfo
+    -> [Api.ChainPoint]                             -- ^ resume points
+    -> IO LCIClient
+newLCIClient info resumePoints = do
+    slotVar <- STM.newTVarIO $ Api.SlotNo 0
+    dataVar <- STM.newTVarIO Map.empty
+
+    let cb = chainSyncCallback slotVar dataVar
+
+    a <- newChainSync info resumePoints cb
+    return $ LCIClient a slotVar dataVar
+
+chainSyncCallback :: STM.TVar Api.SlotNo -> STM.TVar (Map (Api.Hash Api.ScriptData) Api.HashableScriptData) -> ChainSyncCallback
+chainSyncCallback slotVar dataVar (RollForward block@(Api.BlockInMode Api.BabbageEra (Api.Block (Api.BlockHeader slot _ _) _txs)) _tip) =
+    STM.atomically $ do
+        STM.writeTVar slotVar slot
+        STM.modifyTVar' dataVar $ \m ->
+            foldl' (\m' sd -> Map.insert (Api.hashScriptDataBytes sd) sd m') m (blockDatums block)
+
+chainSyncCallback _ _ _ = return ()
+
+-- | Close (destroy) 'LCIClient'.
+closeLCIClient :: LCIClient -> IO ()
+closeLCIClient (LCIClient a _ _) = Async.cancel a
+
+-- | Wait until 'LCIClient' has processed a given slot.
+lciWaitUntilSlot :: LCIClient -> GYSlot -> IO GYSlot
+lciWaitUntilSlot (LCIClient _ slotVar _) (slotToApi -> slot) = STM.atomically $ do
+    slot' <- STM.readTVar slotVar
+    unless (slot' >= slot) STM.retry
+    return (slotFromApi slot')
+
+lookupApiDatum :: LCIClient -> Api.Hash Api.ScriptData -> IO (Maybe Api.HashableScriptData)
+lookupApiDatum (LCIClient _ _ dataVar) h = do
+    m <- STM.readTVarIO dataVar
+    return $ Map.lookup h m
+
+lciLookupDatum :: LCIClient -> GYLookupDatum
+lciLookupDatum c dh = fmap datumFromApi' <$> lookupApiDatum c (datumHashToApi dh)
+
+-- | This is not good 'GeniusYield.Types.Providers.gyGetCurrentSlot' provider as it might lag
```

Comment:
Why is this the case?

Reply:
The chain sync client and the TVar update might add latency.